cli/azd/pkg/pipeline/pipeline.go

```diff
@@ -133,6 +134,10 @@ func (c *ConfigOptions) SecretsAndVars(
 	secrets = maps.Clone(initialSecrets)
 
 	for key, value := range env {
+		if value == "" {
+			// skip empty values
+			continue
+		}
```

Comment:
I don't believe we have enough evidence to force us to skip env variables that are set to an empty string.  Empty string vs env var not being defined can be 2 very different things

Reply:
For CI/CD, an empty string is the same as a non-set variable.
This is because the env var reference is part of the CI definition. For example:

```
- name: Provision Infrastructure
        run: azd provision --no-prompt
        env:
          AZURE_ENV_NAME: ${{ vars.AZURE_ENV_NAME }}
          AZURE_LOCATION: ${{ vars.AZURE_LOCATION }}
          AZURE_SUBSCRIPTION_ID: ${{ vars.AZURE_SUBSCRIPTION_ID }}
```

There's no way to describe that if the value for AZURE_ENV_NAME is not defined as a variable, AZURE_ENV_NAME should not be defined as ENV VAR for the step.  So, as soon as AZURE_ENV_NAME is part of the `env` list, it means it will be defined.

Azure DevOps supports settings conditional-expressions as a way to construct the pipeline definition, but IIRC, the expressions are evaluated before runtime.

In the future, I am planning to use go-templates to provide that functionality, were you can condition when to set env var or anything and azd would construct the definition from there.

Reply:
@vhvb1989 Just to satisfy my curiosity, what happens if we remove this block of code? I would very much favor us doing the "dumb thing" over the "smart thing" here. Perhaps a user means to set this to empty as a placeholder and would very much like to see it in GitHub.

If the individual CI provider rejects the empty string, it could be worthwhile handling this condition within the CI provider. But if we think this is the right thing for all CI providers, we should update the comment to reflect the strong assumptions made here.

Reply:
> what happens if we remove this block of code?

azd returns an error when trying to set the secret or variable with empty string for github.

Keep in mind that, setting the secret or variable in the CI-provider does not mean that it automatically injected for the pipeline. The pipeline definition needs to explicitly create the environment for the entire job or for individual steps, and that's where this value can be mapped.